### PR TITLE
 Implements  Line parameter function. 

### LINKED ISSUE #13515: Address changes 'parameter' finding function
Address the requested changes in PR #11069. The name of the function could be `parameter_value`, but when does submit PR for this issue, we can decide as to what will be a good choice of name.

### PR DESCRIPTION
**Pull request by yashkgp**
Address changes to 'parameter' finding function PR #11069 .Followinfg changes have been made:
1.Changes the name of the function to parameter_value
2.Double backtick point .
3.Makes a testing function for parameter_value
4.Adds a note .
5.Fixes bugs of `parameter_value` function
Fixes #13515

### CODE REVIEW

**Comment by gxyd** [comment]
Thanks for the PR. I am sorry, I am catching a deadline so may not be able to review this in a good way.
I'll make sure someone else reviews your PR.

File: sympy/geometry/line.py [hunk 1]
```diff
@@ -1070,6 +1071,48 @@ def contains(self, other):
             return Point.is_collinear(self.p1, self.p2, other.p1, other.p2)
         return False
 
+    def parameter_value(self, other):
+        """
+        Finds the parameter ``t`` for which
```
Comment (jksuom) [inline sympy/geometry/line.py]: `t` and `1 - t` are usually known as [barycentric coordinates](https://en.wikipedia.org/wiki/Barycentric_coordinate_system) of the point `x`. I think that something like `coordinate` would be a better name for the function.

File: sympy/geometry/line.py [hunk 2]
```diff
@@ -1070,6 +1071,48 @@ def contains(self, other):
...
+        >>> l = Line(p1, p2)
+        >>> l.parameter_value(p3)
+        2
+        """
+        if isinstance(other, Point):
+            if Point.is_collinear(self.p1, self.p2, other):
+                t = Dummy("t")
+                x, y = self.arbitrary_point(t).args
+                if self.p1.x != self.p2.x:
+                    ti = list(solveset(x - other.x, t))[0]
```
Comment (jksuom) [inline sympy/geometry/line.py]: The solution `t` of the equation `x - other.x == 0`, or `self.p1.x + (self.p2.x - self.p1.x)*t == other.x` in expanded form, is easy to write down explicitly. There should be no need to call `solveset`.

File: sympy/geometry/line.py (hunk 1 above)
Comment (yashkgp) [reply sympy/geometry/line.py]: Sure ,will do that.

**Comment by yashkgp** [comment]
Hey ! I have made the changes you asked for.It would be great if you can give it a review.

File: sympy/geometry/line.py [hunk 3]
```diff
@@ -1097,17 +1096,17 @@ def parameter_value(self, other):
...
         >>> p1, p2, p3 = Point(0, 1), Point(2, 3), Point(4, 5)
         >>> l = Line(p1, p2)
-        >>> l.parameter_value(p3)
+        >>> l.coordinate(p3)
         2
         """
         if isinstance(other, Point):
             if Point.is_collinear(self.p1, self.p2, other):
                 t = Dummy("t")
                 x, y = self.arbitrary_point(t).args
```
Comment (jksuom) [inline sympy/geometry/line.py]: These two lines are no more needed.

File: sympy/geometry/line.py (hunk 1 above)
Comment (jksuom) [reply sympy/geometry/line.py]: It looks like the note could be removed.

**Comment by yashkgp** [comment]
Hey ! I have done the requested changes . Can you give it a review please?

File: sympy/geometry/line.py [hunk 4]
```diff
@@ -1070,6 +1070,40 @@ def contains(self, other):
...
+    def coordinate(self, other):
+        """
+        Finds the parameter ``t`` for which
+        ``other == t * self.p1 + (1 - t) self.p2`` is True.
+
+        Raises
+        ======
+
+        ValueError
+            When ``other`` is not contained in the line.
```
Comment (smichr) [inline sympy/geometry/line.py]: I would like to see the NotImplementedError converted to a ValueError. So ValueError would be raised if other is not a Point.

File: sympy/geometry/line.py [hunk 5]
```diff
@@ -1070,6 +1070,40 @@ def contains(self, other):
...
+        Examples
+        ========
+
+        >>> from sympy import Point, Line
+        >>> p1, p2, p3 = Point(0, 1), Point(2, 3), Point(4, 5)
+        >>> l = Line(p1, p2)
+        >>> l.coordinate(p3)
+        2
+        """
+        if isinstance(other, Point):
```
Comment (smichr) [inline sympy/geometry/line.py]: insert `other = Point(other)`

File: sympy/geometry/line.py [hunk 6]
```diff
@@ -1070,6 +1070,40 @@ def contains(self, other):
...
+            if Point.is_collinear(self.p1, self.p2, other):
+                if self.p1.x != self.p2.x:
+                    ti = (other.x - self.p1.x)/(self.p2.x - self.p1.x)
+                else:
+                    ti = (other.y - self.p1.y)/(self.p2.y - self.p1.y)
+                if ti.is_number:
+                    return ti
+            raise ValueError("Point is not collinear.")
+        raise NotImplementedError()
```
Comment (smichr) [inline sympy/geometry/line.py]: convert to `raise ValueError('other should be a Point')`

File: sympy/geometry/tests/test_line.py [hunk 7]
```diff
@@ -286,6 +286,15 @@ def test_contains():
         assert len(w) == 1
 
 
+def test_coordinate():
+    p1, p2, p3 = Point(0, 1), Point(2, 3), Point(3, 4)
+    l = Line(p1, p2)
+    # Line parameter
+    assert l.coordinate(p3) == Rational(3, 2)
```
Comment (smichr) [inline sympy/geometry/tests/test_line.py]: also, `assert l.coordinate(p3.args) == Rational(3, 2)`

File: sympy/geometry/line.py (hunk 5 above)
Comment (smichr) [reply sympy/geometry/line.py]: Actually, perhaps better to say `other = Point._normalize_dimension(Point(other), dim=self.p1.ambient_dimension)` if this should work for a 3-D line.

Comment (yashkgp) [reply sympy/geometry/line.py]: Hey ! @smichr , 
`other = Point._normalize_dimension(Point(other),dim=self.p1.ambient_dimension) ` is giving a `Value error` .I tried many things but was unable to remove the error. It would be great if you can tell me where I am going wrong . The error that I am getting is : 
```
  File "/home/yash_jain/Desktop/env/sympy/sympy/geometry/tests/test_line.py", line 293, in test_coordinate
    assert l.coordinate(p3) == Rational(3, 2)
  File "/home/yash_jain/Desktop/env/sympy/sympy/geometry/line.py", line 1098, in coordinate
    if Point.is_collinear(self.p1, self.p2, other):
  File "/home/yash_jain/Desktop/env/sympy/sympy/geometry/point.py", line 538, in is_collinear
    points = Point._normalize_dimension(*[Point(i) for i in points])
  File "/home/yash_jain/Desktop/env/sympy/sympy/geometry/point.py", line 538, in <listcomp>
    points = Point._normalize_dimension(*[Point(i) for i in points])
  File "/home/yash_jain/Desktop/env/sympy/sympy/geometry/point.py", line 141, in __new__
    keyword `dim` > 1.'''))
ValueError: 
Point requires 2 or more coordinates or keyword `dim` > 1.
```

File: sympy/geometry/tests/test_line.py (hunk 7 above)
Comment (yashkgp) [reply sympy/geometry/tests/test_line.py]: Hey,
This line of code is giving a `Value Error: Other should be a point . `

**Comment by yashkgp** [comment]
Hey ! I have made the requested changes . Can you give it a review?

File: sympy/geometry/line.py [hunk 8]
```diff
@@ -1070,6 +1070,40 @@ def contains(self, other):
...
+        2
+        """
+        if isinstance(other, Point):
+            other=Point(other)
+            if Point.is_collinear(self.p1, self.p2, other):
+                if self.p1.x != self.p2.x:
+                    ti = (other.x - self.p1.x)/(self.p2.x - self.p1.x)
+                else:
+                    ti = (other.y - self.p1.y)/(self.p2.y - self.p1.y)
+                if ti.is_number:
```
Comment (jksuom) [inline sympy/geometry/line.py]: Is this necessary?

Comment (yashkgp) [reply sympy/geometry/line.py]: Its for the case if `self.p2.y ==self.p1.y:`, then the else statement will explode . We can remove this if statement and convert the else statement to `elif self.p2.y !=self.p1.y:`  This will serve the same purpose.

Comment (jksuom) [reply sympy/geometry/line.py]: > convert the else statement to elif self.p2.y !=self.p1.y:

I think this is the proper way. It will allow the use of symbolic values for coordinates.

File: sympy/geometry/line.py [hunk 9]
```diff
@@ -1097,10 +1097,9 @@ def coordinate(self, other):
             if Point.is_collinear(self.p1, self.p2, other):
                 if self.p1.x != self.p2.x:
                     ti = (other.x - self.p1.x)/(self.p2.x - self.p1.x)
-                else:
+                elif self.p1.y != self.p2.y:
                     ti = (other.y - self.p1.y)/(self.p2.y - self.p1.y)
-                if ti.is_number:
-                    return ti
+                return ti
```
Comment (jksuom) [inline sympy/geometry/line.py]: If the two points coincide, `ti` will be undefined. It is probably better to replace each `ti =` by `return` and omit this line.

File: sympy/geometry/line.py [hunk 10]
```diff
@@ -1096,10 +1096,9 @@ def coordinate(self, other):
             other=Point(other)
             if Point.is_collinear(self.p1, self.p2, other):
                 if self.p1.x != self.p2.x:
-                    ti = (other.x - self.p1.x)/(self.p2.x - self.p1.x)
+                    return (other.x - self.p1.x)/(self.p2.x - self.p1.x)
                 elif self.p1.y != self.p2.y:
-                    ti = (other.y - self.p1.y)/(self.p2.y - self.p1.y)
-                return ti
+                    return (other.y - self.p1.y)/(self.p2.y - self.p1.y)
```
Comment (jksuom) [inline sympy/geometry/line.py]: Should there be
``` 
   else:
        ValueError(...)
```
after this line to indicate that the endpoints coincide?

File: sympy/geometry/line.py (hunk 5 above)
Comment (smichr) [reply sympy/geometry/line.py]: How about `other = self.p1._normalize_dimension(Point(other))`?

Comment (smichr) [reply sympy/geometry/line.py]: And that should be the first line of the function.

**Comment by yashkgp** [comment]
Hey @jksuom I have made the requested changes . Could you give it a review?

File: sympy/geometry/line.py [hunk 11]
```diff
@@ -1070,6 +1070,39 @@ def contains(self, other):
             return Point.is_collinear(self.p1, self.p2, other.p1, other.p2)
         return False
 
+    def coordinate(self, other):
```
Comment (smichr) [inline sympy/geometry/line.py]: Would it make sense to make the arg list be `self, x, y=None` and then `line.coordinate(3)` would give `t` where the Point on the line whose x value is 3 would be returned? (It would raise an error if the line was vertical and didn't go through `x=3`. `line.coordinate(3, 4)` would behave as it does now. And `line.coordinate(y=4)` would behave the same as the `x=` case but raise an error for a horizontal line not passing through 4.

Comment (jksuom) [reply sympy/geometry/line.py]: If I understand you correctly that would involve essentially two steps: 1) from plane coordinate x to point P on the line, 2) from P to its affine line coordinate. Perhaps there could be another function for step 1) alone?

Comment (smichr) [reply sympy/geometry/line.py]: Yes, I see your point. I didn't appreciate the difference in dealing with 2D vs 3D lines. It need not be part of this PR. I'll leave the final commit to you on this one.

File: sympy/geometry/line.py [hunk 12]
```diff
@@ -1070,6 +1070,39 @@ def contains(self, other):
...
+        2
+        """
+        if not isinstance(other, GeometryEntity):
+            other = Point(other, dim=self.ambient_dimension)
+        if isinstance(other, Point):
+            if Point.is_collinear(self.p1, self.p2, other):
+                if self.p1.x != self.p2.x:
+                    return (other.x - self.p1.x)/(self.p2.x - self.p1.x)
+                elif self.p1.y != self.p2.y:
+                    return (other.y - self.p1.y)/(self.p2.y - self.p1.y)
```
Comment (jksuom) [inline sympy/geometry/line.py]: Should we also think of the dim 3 case? Then something like this should probably be added:

    elif dim == 3:
        return (other.z - self.p1.z)/(self.p2.z - self.p1.z)

**Comment by smichr** [comment]
Two comments:

1. the title of this PR suggest the new method is named 'parameter' and that makes more sense to me than coordinate; maybe `affine_pt` would be better.
1. Is there any reason not to just let solve handle all the details?
```
def parameter(self, other):
    from sympy import solve
    if not isinstance(other, GeometryEntity):
        other = Point(other, dim=self.ambient_dimension)
    if not isinstance(other, Point):
        raise ValueError
    k = Dummy()
    sol = solve(self.arbitrary_point(k) - other, dict=True) or [{}]
    v = sol[0].get(k, None)
    if v is None:
        raise ValueError('given point is not on the line')
    return v
```
Examples
```
>>> L = Line((1,2,3),(2,3,4))
>>> parameter(Line((1,2,3),(2,3,4)),(x,6,z))
4
>>> parameter(Line((1,2,3),(2,3,4)),(x,y,z))
z - 3
>>> parameter(Line((1,2,3),(2,3,4)),(0,y,z))
-1
>>> parameter(Line((1,2,3),(2,3,4)),(x,0,z))
-2
>>> parameter(Line((1,2,3),(2,3,4)),(x,y,0))
-3
>>> parameter(Line((1,2,3),(2,3,4)),(3,y,0))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 11, in parameter
ValueError: given point is not on the line
>>> parameter(Line((1,2,3),(2,3,4)),(3,y,5))
2
>>> parameter(Line((1,2,3),(2,3,4)),(x,y,5))
2
>>> parameter(Line((1,2,3),(2,3,4)),(3,4,5))
2
```

**Comment by jksuom** [comment]
I think that the parameters defining the position of a point in a coordinate system are usually called coordinates. The natural coordinate system on a line given by two points is the affine coordinate system, which  is defined by value 0 at one point and value 1 at the other one.

**Comment by yashkgp** [comment]
I was also thinking the same , `solve` will make the things much easier even for 3D lines . Should a send a commit for the same?

**Comment by smichr** [comment]
> The natural coordinate system on a line given by two points is the affine coordinate system, which is defined by value 0 at one point and value 1 at the other one.

The output of the function under consideration is then `affine_coordinate` which is not a lot of fun to type but it's clear. :-)

**Comment by smichr** [comment]
> send a commit for the same

feel free to modify what I've posted, if you want

**Comment by yashkgp** [comment]
Hey! I tried involving  `solveset` in the function , but I think its of no use, as we have to still deal with the three cases differently. I made a function something like this : 
``` python
    def paramet(self, other):
        from sympy.solvers.solveset import solveset
        from sympy import Symbol,pprint,Eq,Dummy
        if not isinstance(other, GeometryEntity):
            other = Point(other, dim=self.ambient_dimension)
        if isinstance(other, Point):
            p = Symbol('p')
            if Point.is_collinear(self.p1, self.p2, other):
                if self.p1.x != self.p2.x:
                    sol = solveset(self.arbitrary_point(p).x - other.x,p )
                elif self.p1.y != self.p2.y:
                    sol = solveset(self.arbitrary_point(p).y - other.y,p )
                elif self.p1.ambient_dimension == 3:
                    sol = solveset(self.arbitrary_point(p).z - other.z,p )
                return sol
            raise ValueError("Point is not collinear.")
        raise ValueError('Other should be a Point')
```
Should I send a PR anyways ?

**Comment by smichr** [comment]
Does the code I posted [here](https://github.com/sympy/sympy/pull/13539#issuecomment-350046130) have the same problem?

**Comment by smichr** [comment]
> p = Symbol('p')

You'll want to use a Dummy in case they already have a symbol p in defining the line.

**Comment by yashkgp** [comment]
@smichr Yes , your code was having the same problem . Using their symbol raised a `ValueError` that why I made a different symbol

**Comment by smichr** [comment]
Could you please post an example that fails.

**Comment by yashkgp** [comment]
Using their symbol which is `t` I ran a test : 
```python
L = Line((1,2,3),(1,4,6))
print (L.paramet((1,6,9)))
print (L.coordinate((1,6,9)))
```
It gave  output
```
EmptySet()
2
```
Before It was giving a Error because I haven't initiated  `t` but now it giving a `EmptySet()` because its treating both the `t` 's differently . This the code that i have used : 
```python
    def paramet(self, other):
        from sympy.solvers.solveset import solveset
        from sympy import Symbol,pprint,Eq,Dummy
        if not isinstance(other, GeometryEntity):
            other = Point(other, dim=self.ambient_dimension)
        if isinstance(other, Point):
            t = Dummy('t')
            if Point.is_collinear(self.p1, self.p2, other):
                if self.p1.x != self.p2.x:
                    sol = solveset(self.arbitrary_point().x - other.x,t )
                elif self.p1.y != self.p2.y:
                    sol = solveset(self.arbitrary_point().y - other.y,t )
                elif self.p1.ambient_dimension == 3:
                    sol = solveset(self.arbitrary_point().z - other.z, t )
                return sol
            raise ValueError("Point is not collinear.")
        raise ValueError('Other should be a Point')
```
where `t` is the symbol in arbitrary point

**Comment by smichr** [comment]
The routine I gave seems to work ok. Was there are case you found where it didn't work?

```python
>>> from sympy.geometry.entity import GeometryEntity
>>> def parameter(self, other):
...     from sympy import solve
...     if not isinstance(other, GeometryEntity):
...         other = Point(other, dim=self.ambient_dimension)
...     if not isinstance(other, Point):
...         raise ValueError
...     k = Dummy()
...     sol = solve(self.arbitrary_point(k) - other, dict=True) or [{}]
...     v = sol[0].get(k, None)
...     if v is None:
...         raise ValueError('given point is not on the line')
...     return v
...
>>> Line.parameter = parameter
>>> L = Line((1,2,3),(1,4,6))
>>> L.parameter((1,6,9))
2
```

**Comment by smichr** [comment]
btw, you aren't telling arbitrary point to use your symbol:

```
    self.arbitrary_point().x - other.x, t 
should be
    self.arbitrary_point(t).x - other.x, t
                         ^
```

File: sympy/geometry/line.py (hunk 12 above)
Comment (smichr) [reply sympy/geometry/line.py]: +1. It should handle a 3rd dimension and raise an error for higher dimensions.

File: sympy/physics/units/definitions.py [hunk 13]
```diff
@@ -137,7 +137,7 @@
 amu = amus = atomic_mass_unit = atomic_mass_constant = Quantity("atomic_mass_constant", mass, 1.660539040e-24*gram)
 # Molar gas constant
 R = molar_gas_constant = Quantity("molar_gas_constant", energy/(temperature * amount_of_substance),
-                                  8.3144598*joule/kelvin/mol, abbrev="R")
+                                  avogadro_number*boltzmann, abbrev="R")
```
Comment (smichr) [inline sympy/physics/units/definitions.py]: Where are these changes coming from? (Let's keep these as a separate commit if they are yours.)

**Comment by smichr** [comment]
Let's put [my routine](https://github.com/sympy/sympy/pull/13539#issuecomment-350505163) (or similar) at the top level of geometry: any object that defines `arbitrary_point` will be able to use that routine to return the corresponding affine coordinate. I am now -1 to this routine. @yashkgp , do you think you can make the change that I suggested? I *think* it would go into GeometricEntity. An error should be raised if the self does not have the `arbitrary_point` attribute.
